src/realm/dictionary.cpp

```diff
@@ -433,16 +433,8 @@ void Dictionary::insert_collection(const PathElement& path_elem, CollectionType
     if (dict_or_list == CollectionType::Set) {
         throw IllegalOperation("Set nested in Dictionary is not supported");
     }
-
     check_level();
-    ensure_created();
-    Mixed new_val(0, dict_or_list);
-    auto old_val = try_get(path_elem.get_key());
-    if (!old_val || *old_val != new_val) {
-        m_values->ensure_keys();
-        auto [it, inserted] = insert(path_elem.get_key(), new_val);
-        set_key(*m_values, it.index());
-    }
+    insert(path_elem.get_key(), Mixed(0, dict_or_list));
```

Comment:
isn't it cleaner to keep invoking `set_key` from here and not in `insert`? we can use the value returned by insert to decide if it's needed. Edit: I guess the idea is to be able to also insert/set a collection through a Mixed value, which probably didn't work before.

Reply:
The idea is just to reuse the code in Dictionary::insert. We can't use the return value to determine if we should call `set_key`. The return value tells if a new entry was added - not if we have updated a value.

Reply:
Right. but doesn't that mean that if a key was not inserted it was updated instead? anyway, I get why we want that logic in insert (so we can insert a collection through set(Mixed) too)